Environments article overhaul (#36022)
dotnet/AspNetCore.Docs: aspnetcore/fundamentals/environments.md

* [Caching](xref:performance/caching/memory) is enabled.
* Client-side resources are bundled, minified, and potentially served from a CDN.
* Diagnostic error pages are disabled.
* Friendly error pages are enabled.
* Production logging and monitoring are enabled. For example, logging is enabled for [Azure Application Insights](/azure/application-insights/app-insights-asp-net-core).

The last environment setting read by the app determines the app's environment. The app's environment can't be changed while the app is running.

## Logging

Output in the command shell of a running app at startup indicates the app's environment. In the following example, the app is running in the Staging environment:

```dotnetcli
info: Microsoft.Hosting.Lifetime[0]
      Hosting environment: Staging
```

## Environment variables that determine the runtime environment

To determine the runtime environment, ASP.NET Core reads from the following environment variables:

:::moniker range=">= aspnetcore-7.0"

* [`DOTNET_ENVIRONMENT`](xref:fundamentals/configuration/index#default-host-configuration)

> * `ASPNETCORE_ENVIRONMENT` when the <xref:Microsoft.AspNetCore.Builder.WebApplication.CreateBuilder%2A?displayProperty=nameWithType> method is called. The ASP.NET Core web app project templates call `WebApplication.CreateBuilder`. The `DOTNET_ENVIRONMENT` value overrides `ASPNETCORE_ENVIRONMENT` when <xref:Microsoft.AspNetCore.Builder.WebApplicationBuilder> is used. For other hosts, such as <xref:Microsoft.Extensions.Hosting.GenericHostBuilderExtensions.ConfigureWebHostDefaults%2A> and <xref:Microsoft.AspNetCore.WebHost.CreateDefaultBuilder%2A?displayProperty=nameWithType>, `ASPNETCORE_ENVIRONMENT` has higher precedence.

Reviewer: Can we simplify this language?

```suggestion
* `ASPNETCORE_ENVIRONMENT`

When using <xref:Microsoft.AspNetCore.Builder.WebApplication>, the `DOTNET_ENVIRONMENT` value take precedence over `ASPNETCORE_ENVIRONMENT`. When using <xref:Microsoft.AspNetCore.Builder.WebHost>, `ASPNETCORE_ENVIRONEMNT` takes precedence.
```

:::moniker-end

:::moniker range=">= aspnetcore-6.0 < aspnetcore-7.0"

* [`DOTNET_ENVIRONMENT`](xref:fundamentals/configuration/index#default-host-configuration)
* `ASPNETCORE_ENVIRONMENT` when the <xref:Microsoft.AspNetCore.Builder.WebApplication.CreateBuilder%2A?displayProperty=nameWithType> method is called. The ASP.NET Core web app project templates call `WebApplication.CreateBuilder`. The `ASPNETCORE_ENVIRONMENT` value overrides `DOTNET_ENVIRONMENT`.

> :::moniker-end

Reviewer: Isn't this saying pretty much the same thing as the >= aspnetcore-7.0 version above?

Author: The override is reversed (at least it says so, as I never kept this tidbit in my head) ...

* .NET 7: `DOTNET_ENVIRONMENT` value overrides `ASPNETCORE_ENVIRONMENT`
* .NET 6: `ASPNETCORE_ENVIRONMENT` value overrides `DOTNET_ENVIRONMENT`

Reviewer: Ah, I need to learn to read more carefully! Thanks!

:::moniker range="< aspnetcore-6.0"

* [`DOTNET_ENVIRONMENT`](xref:fundamentals/configuration/index#default-host-configuration)
* `ASPNETCORE_ENVIRONMENT` when <xref:Microsoft.Extensions.Hosting.GenericHostBuilderExtensions.ConfigureWebHostDefaults%2A> is called. The ASP.NET Core web app project templates call `ConfigureWebHostDefaults`. The `ASPNETCORE_ENVIRONMENT` value overrides `DOTNET_ENVIRONMENT`.

:::moniker-end

If the `DOTNET_ENVIRONMENT` and `ASPNETCORE_ENVIRONMENT` environment variables aren't set, the Production environment is the default environment.

On Windows and macOS, environment variable names aren't case-sensitive. Linux environment variables are case-sensitive.

## Control code execution by environment

:::moniker range=">= aspnetcore-6.0"

Use <xref:Microsoft.AspNetCore.Builder.WebApplicationBuilder.Environment?displayProperty=nameWithType> or <xref:Microsoft.AspNetCore.Builder.WebApplication.Environment?displayProperty=nameWithType> to conditionally add services or middleware depending on the current environment.

The following code in the app's `Program` file:

> * Uses <xref:Microsoft.AspNetCore.Builder.WebApplication.Environment?displayProperty=nameWithType> to distinguish the environment, specifically when the value of `ASPNETCORE_ENVIRONMENT` is anything other than `Development`. This approach is useful when the app only requires adjusting the `Program` file for a few environments with minimal code differences per environment. When many code differences exist per environment, consider using one of the [environment-based `Startup` class approaches](#environment-based-startup-class-and-methods), which is covered later in this article.

Reviewer: I don't think we want to recommend the older `Startup` class pattern.

```suggestion
* Uses <xref:Microsoft.AspNetCore.Builder.WebApplication.Environment?displayProperty=nameWithType> to distinguish the environment.
```

```csharp
public void Configure(IApplicationBuilder app, IWebHostEnvironment env)
{
    if (env.IsDevelopment())
    {
        app.UseDeveloperExceptionPage();
    }

    if (env.IsProduction() || env.IsStaging() || env.IsEnvironment("Testing"))
    {
        app.UseExceptionHandler("/Error");
    }

    ...
}
```

The preceding example checks the current environment while building the request pipeline. To check the current environment in `Startup.ConfigureServices` while configuring services, inject <xref:Microsoft.AspNetCore.Hosting.IWebHostEnvironment> into the `Startup` class instead of injecting it into `Startup.Configure` and use the injected service to determine the environment in `Startup.ConfigureServices` and `Startup.Configure`.

:::moniker-end

Within the app, the <xref:Microsoft.Extensions.Hosting.IHostEnvironment> provides general information about the app's hosting environment, and the <xref:Microsoft.Extensions.Hosting.IHostEnvironment.EnvironmentName?displayProperty=nameWithType> property indicates the app's current environment.

> ## Control rendered content in Razor Pages and MVC

Author: Yes ... but again leading to the question of why ***all*** Blazor *Fundamentals* node content isn't going to move over here? Specifically in this case, all of Blazor's *Environments* article coverage ...

https://learn.microsoft.com/en-us/aspnet/core/blazor/fundamentals/environments?view=aspnetcore-9.0

I mean to really lead with Blazor, it wouldn't be squirreled back in what is the equivalent of the "UI stack" coverage for MVC, RP, or Blazor.

> ## Control rendered content in Razor Pages and MVC

Reviewer: > Yes ... but again leading to the question of why all Blazor Fundamentals node content isn't going to move over here? Specifically in this case, all of Blazor's Environments article coverage ...

Yup, I think that's the right thing to do, but we need to do so tastefully. The fundamentals of ASP.NET Core should feel simple and broadly applicable. As we merge the Blazor fundamentals into the ASP.NET Core fundamentals we should be careful to remove any content duplication and to avoid excessive Blazor specific details. I don't want to overwhelm the ASP.NET Core fundamentals docs with a bunch of Blazor content that's specific to narrow scenarios. So, let's be judicious about how we do the merge. If half of the merged Environments doc ends up being about Blazor WebAssembly specific concerns, that's probably not ideal.

  "commandName": "Project",
  "dotnetRunMessages": true,
  "launchBrowser": true,
  "applicationUrl": "https://localhost:7205",
  "environmentVariables": {
    "ASPNETCORE_ENVIRONMENT": "Staging"
  }
}
```

In Visual Studio, there are two approaches for setting the environment via launch profiles:

* Press <kbd>Alt</kbd>+<kbd>Enter</kbd> or select **Properties** after right-clicking the project in **Solution Explorer**. Select **Debug** > **General**, followed by selecting the **Open debug launch profiles UI** link.

* With the project selected in **Solution Explorer**, select **{PROJECT NAME} Debug Properties** from the **Debug** menu, where the `{PROJECT NAME}` placeholder is a project name.

The preceding approaches open the **Launch Profiles** dialog where you can edit the environment variable settings in the `launchSettings.json` file. Changes made to project profiles may not take effect until the web server is restarted. Kestrel must be restarted before it can detect changes made to its environment.

Profiles can be selected in the Visual Studio UI next to the Start button (►).

When a solution contains multiple projects, only set the environment for the startup project.

> Alternatively, use the [`dotnet run`](/dotnet/core/tools/dotnet-run) command with the [`-lp|--launch-profile` option](/dotnet/core/tools/dotnet-run#options) set to the profile's name. *This approach only supports Kestrel profiles.*

Reviewer: ```suggestion
Alternatively, use the [`dotnet run`](/dotnet/core/tools/dotnet-run) command with the [`-lp|--launch-profile` option](/dotnet/core/tools/dotnet-run#options) set to the profile's name. *This approach only supports launch profiles based on the `Project` command.*
```

```dotnetcli
dotnet run -lp "https"
```

When using [Visual Studio Code](https://code.visualstudio.com/), set the `ASPNETCORE_ENVIRONMENT` environment variable in the `.vscode/launch.json` in the `env` section, along with any other environment variables set in the section:

```json
"env": {
    "ASPNETCORE_ENVIRONMENT": "Staging",
    ...
},
```

> The `.vscode/launch.json` file is only used by Visual Studio Code.

Reviewer: Using *launch.json* isn't necessary with C# Dev Kit. C# Dev Kit will pick up the available launch profiles from your *launchSettings.json* file.

> ## Set the environment with an environment variable

Reviewer: I really don't think we need to be documenting how to set an environment variable on various platforms, like Windows, macOS, and Linux. I'm ok keeping some pointers on how to do it for App Service and IIS, although I'd prefer those pointers just be links to the docs for the corresponding service to tech.

```shell
docker run --env-file ./env_variables.env aspnet_core_image
```

An environment set at runtime with `docker run` overrides an environment set by the Dockerfile.

## Set the environment in the app's startup code

:::moniker range=">= aspnetcore-6.0"

To set the environment in code, use <xref:Microsoft.AspNetCore.Builder.WebApplicationOptions.EnvironmentName?displayProperty=nameWithType> when creating <xref:Microsoft.AspNetCore.Builder.WebApplicationBuilder>, as shown in the following example:

```csharp
var builder = WebApplication.CreateBuilder(new WebApplicationOptions
{
    EnvironmentName = Environments.Staging
}); 
```

:::moniker-end

:::moniker range="< aspnetcore-6.0"

Call <xref:Microsoft.Extensions.Hosting.HostingHostBuilderExtensions.UseEnvironment%2A> when building the host. For more information, see <xref:fundamentals/host/generic-host#environmentname>.

:::moniker-end

## Load configuration by environment

To load configuration by environment, see <xref:fundamentals/configuration/index#json-configuration-provider>.

## Environment-based `Startup` class and methods

> Since the release of .NET 6, the ASP.NET Core project templates place startup code for service configuration and request pipeline processing in the `Program` file. Use of a `Startup` class (`Startup.cs`) with [`Configure`](xref:Microsoft.AspNetCore.Hosting.StartupBase.Configure%2A) and [`ConfigureServices`](xref:Microsoft.AspNetCore.Hosting.StartupBase.ConfigureServices%2A) methods was required before the release of .NET 6 but remains supported in ASP.NET Core. You can use the guidance in this section to control code execution by environment using a `Startup` class in any release of ASP.NET Core when a large number of code changes are required for each environment in use by an app.

Reviewer: I prefer to avoid history lessons in our docs. Let's just document our features and guidance.

```suggestion
```

> Except for the [`Startup` class conventions](#startup-class-conventions) approach, the `Program` class uses the following general approach to identify the `Startup` class, where the `Startup` class is the type for the <xref:Microsoft.AspNetCore.Hosting.WebHostBuilderExtensions.UseStartup%2A?displayProperty=nameWithType> method:

Reviewer: It is odd that we no longer document the old Startup class pattern in our app startup doc. We should probably fix that. But I think we can assume the reader is familiar with it if they care to read this section.

> Except for the [`Startup` class conventions](#startup-class-conventions) approach, the `Program` class uses the following general approach to identify the `Startup` class, where the `Startup` class is the type for the <xref:Microsoft.AspNetCore.Hosting.WebHostBuilderExtensions.UseStartup%2A?displayProperty=nameWithType> method:

Reviewer: It is odd that we no longer document the old Startup class pattern in our app startup doc. We should probably fix that. But I think we can assume the reader is familiar with it if they care to read this section.

```suggestion
```

Author: I did have this here for visibility to those who have never seen it or forgotten about it.

```csharp
public class Program
{
    public static void Main(string[] args)
    {
        CreateHostBuilder(args).Build().Run();
    }

>     public static IHostBuilder CreateHostBuilder(string[] args) =>
>         Host.CreateDefaultBuilder(args)
>             .ConfigureWebHostDefaults(webBuilder =>
>             {
>                 webBuilder.UseStartup<Startup>();
>             });
> }

Reviewer: ````suggestion
````

```

> ### Inject `IWebHostEnvironment` into the `Startup` class

Reviewer: ```suggestion
```

> Inject <xref:Microsoft.AspNetCore.Hosting.IWebHostEnvironment> into the `Startup` constructor to control code execution. This approach is useful when the app requires configuring startup code for only a few environments with minimal code differences per environment.
> 
> In the following example, the environment is held in the `_env` field and controls code execution based on the app's environment:
> 
> ```csharp
> public class Startup
> {
>     private readonly IWebHostEnvironment _env;
> 
>     public Startup(IWebHostEnvironment env)
>     {
>         _env = env;
>     }
> 
>     public void ConfigureServices(IServiceCollection services)
>     {
>         if (_env.IsDevelopment())
>         {
>             ...
>         }
>         else if (_env.IsStaging())
>         {
>             ...
>         }
>         else
>         {
>             ...
>         }
>     }
> 
>     public void Configure(IApplicationBuilder app)
>     {
>         if (_env.IsDevelopment())
>         {
>             ...
>         }
>         else
>         {
>             ...
>         }
> 
>         ...
>     }
> }
> ```

Reviewer: The primary constructor version is simpler and more idiomatic with modern C#, so let's just use that and have one sample.

````suggestion
Inject <xref:Microsoft.AspNetCore.Hosting.IWebHostEnvironment> into the `Startup` constructor to access the current environment during startup configuration.
```

Reviewer: We don't have to show the primary constructor version. We could just show a normal constructor as our single sample. Thie startup pattern is an older pattern that is eventually going away, so sticking to the older pattern kinda makes sense. But I definitely don't want to show two version of the same sample using different C# features.

> The preceding example is simplified with the use of [primary constructors](/dotnet/csharp/whats-new/tutorials/primary-constructors) in C# 12 (.NET 8) or later:

Reviewer: ```suggestion
```

```csharp
public class Startup(IWebHostEnvironment env)
{
    public void ConfigureServices(IServiceCollection services)
    {
        if (env.IsDevelopment())
        {
            ...
        }
        else if (env.IsStaging())
        {
            ...
        }
        else
        {
            ...
        }
    }

    public void Configure(IApplicationBuilder app)
    {
        if (env.IsDevelopment())
        {
            ...
        }
        else
        {
            ...
        }

        ...
    }
}
```

> ### `Startup` class conventions

Reviewer: ```suggestion
## `Startup` class per environment convention
```

An app can define multiple `Startup` classes for different environments with the naming convention `Startup{EnvironmentName}` class, where the `{ENVIRONMENT NAME}` placeholder is the environment name.

> The class whose name suffix matches the current environment is prioritized. If a matching `Startup{EnvironmentName}` class isn't found, the `Startup` class is used. This approach is useful when the app requires configuring startup for several environments with many code differences per environment.

Reviewer: ```suggestion
The class whose name suffix matches the current environment is prioritized. If a matching `Startup{EnvironmentName}` class isn't found, the `Startup` class is used.
```

To implement environment-based `Startup` classes, create as many `Startup{EnvironmentName}` classes as needed and a fallback `Startup` class:

```csharp
public class StartupDevelopment
{
    ...
}

public class StartupProduction
{
    ...
}

public class Startup
{
    ...
}
```

Where the host builder is created, call <xref:Microsoft.AspNetCore.Hosting.HostingAbstractionsWebHostBuilderExtensions.UseStartup%2A?displayProperty=nameWithType>, which accepts an assembly name to load the correct `Startup` class:

```csharp
public static IHostBuilder CreateHostBuilder(string[] args)
{
    var assemblyName = typeof(Startup).GetTypeInfo().Assembly.FullName;

    return Host.CreateDefaultBuilder(args)
        .ConfigureWebHostDefaults(webBuilder =>
        {
            webBuilder.UseStartup(assemblyName);
        });
}
```

> ### `Startup` class method conventions

Reviewer: ```suggestion
### `Startup` class method per environment convention
```

> The `Configure` and `ConfigureServices` methods support environment-specific versions of the form `Configure{ENVIRONMENT NAME}` and `Configure{ENVIRONMENT NAME}Services`, where the `{ENVIRONMENT NAME}` placeholder is the environment name. If a matching environment name isn't found for the named methods, the `ConfigureServices` or `Configure` method is used, respectively. This approach is useful when the app requires configuring startup for several environments with many code differences per environment:

Reviewer: ```suggestion
The `Configure` and `ConfigureServices` methods support environment-specific versions of the form `Configure{ENVIRONMENT NAME}` and `Configure{ENVIRONMENT NAME}Services`, where the `{ENVIRONMENT NAME}` placeholder is the environment name. If a matching environment name isn't found for the named methods, the `ConfigureServices` or `Configure` method is used, respectively.
```
